File: Easy/697 - DegreeOfAnArray/solution.ts
// Question 697. Degree of an Array (https://leetcode-cn.com/problems/degree-of-an-array/)

/**
 * 执行用时：112 ms, 在所有 TypeScript 提交中击败了 44.44% 的用户
 * 内存消耗：47.6 MB, 在所有 TypeScript 提交中击败了 11.11% 的用户
 *
 * @param {number[]} nums
 * @returns {number}
 */
function findShortestSubArray(nums: number[]): number {
  const elementsMapping: Map<number, [number, number, number]> = new Map<
    number,
    [number, number, number]
  >();
  const N: number = nums.length;
  let maxFrequency: number = 0;

  // Create a mapping for elements.
  for (let i = 0; i < N; i++) {
    let element = nums[i];
    if (elementsMapping.has(element)) {
      let [startIndex, endIndex, frequency] = elementsMapping.get(element) as [
        number,
        number,
        number
      ];
      frequency++;
      elementsMapping.set(element, [
        startIndex,
        Math.max(endIndex, i),
        frequency,
      ]);
      maxFrequency = Math.max(frequency, maxFrequency);
    } else {
      elementsMapping.set(element, [i, i, 1]);
      maxFrequency = Math.max(1, maxFrequency);
    }
  }

  let minLength: number = N;
  elementsMapping.forEach(([startIndex, endIndex, frequency], key) => {
    // console.log(`key = ${key}, [${startIndex}, ${endIndex}, ${frequency}]`);
    if (frequency == maxFrequency) {
      minLength = Math.min(minLength, endIndex - startIndex + 1);
    }
  });

  return minLength;
}

File: Easy/697 - DegreeOfAnArray/Solution.cs
// Question 697. Degree of an Array (https://leetcode-cn.com/problems/degree-of-an-array/)

/// <summary>
/// 执行用时：140 ms, 在所有 C# 提交中击败了 100.00% 的用户
/// 内存消耗：32.1 MB, 在所有 C# 提交中击败了 76.00% 的用户
/// </summary>
public class Solution 
{
    public int FindShortestSubArray(int[] nums) 
    {
        Dictionary<int, (int, int, int)> elementsMapping = new Dictionary<int, (int, int, int)>();
        int N = nums.Count();
        int maxFrequency = 0;

        // Create a mapping for elements.
        for (int i = 0; i < N; i++)
        {
            int element = nums[i];
            if (elementsMapping.ContainsKey(element))
            {
                (int startIndex, int endIndex, int frequency) = elementsMapping[element];
                frequency++;
                elementsMapping[element] = (startIndex, Math.Max(endIndex, i), frequency);
                maxFrequency = Math.Max(frequency, maxFrequency);
            }
            else
            {
                elementsMapping[element] = (i, i, 1);
                maxFrequency = Math.Max(1, maxFrequency);
            }
        }

        int minLength = N;
        foreach ((int startIndex, int endIndex, int frequency) in elementsMapping.Values)
        {
            if (frequency == maxFrequency)
            {
                minLength = Math.Min(minLength, endIndex - startIndex + 1);
            }
        }

        return minLength;
    }
}